sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { bcs } from '../bcs/index.js';
+import type { SuiClient } from '../client/client.js';
+import type { ExecuteTransactionBlockParams } from '../client/index.js';
+import type { Signer } from '../cryptography/keypair.js';
+import { normalizeSuiAddress } from '../utils/sui-types.js';
+import type { OpenMoveTypeSignature } from './blockData/v2.js';
+import type { TransactionBlock } from './TransactionBlock.js';
+import type { TransactionBlockDataResolverPlugin } from './TransactionBlockDataResolver.js';
+
+export interface AsyncCache<T> {
+	get(key: string): Promise<T | null>;
+	set(key: string, value: T): Promise<void>;
```

Comment:
It's often useful to know what we're storing so that we can control the ttl. E.g. shared objects can get cached forever, never cache bust, but owned objects do.

Reply:
Thats a good point.  Does something like `set<T extends keyof CacheTypes>(type: T, key: string, value: CacheTypes[T])` make sense?

Reply:
Yeah I think works well.